lib/translation.ts

```diff
@@ -0,0 +1,339 @@
+	const defaultOptions = {
+		escape: true,
+		sanitize: true,
+	}
+	const allOptions = Object.assign({}, defaultOptions, options || {})
+
+	const identity = (value) => value
+	const optSanitize = allOptions.sanitize ? DOMPurify.sanitize : identity
+	const optEscape = allOptions.escape ? escapeHTML : identity
+
+	// TODO: cache this function to avoid inline recreation
+	// of the same function over and over again in case
+	// translate() is used in a loop
+	const _build = (text: string, vars?: Record<string, any>, number?: number) => {
+		return text
+			.replace(/%n/g, '' + number)
+			.replace(/{([^{}]*)}/g, (match, key) => {
+				if (vars === undefined || !(key in vars)) return optSanitize(match)
+
+				const r = vars[key]
+				if (typeof r === 'string' || typeof r === 'number') {
+					return optSanitize(optEscape(r))
+				} else {
+					return optSanitize(match)
+				}
+			}
+			)
+	}
+
+	const bundle = getAppTranslations(app)
+	const translation = bundle.translations[text] || text
+
+	if (typeof vars === 'object' || number !== undefined) {
+		return optSanitize(_build(translation, vars, number))
+	} else {
+		return optSanitize(translation)
+	}
+}
+
+/**
+ * Translate a plural string
+ *
+ * @param {string} app the id of the app for which to translate the string
+ * @param {string} textSingular the string to translate for exactly one object
+ * @param {string} textPlural the string to translate for n objects
+ * @param {number} number number to determine whether to use singular or plural
+ * @param {Object} vars of placeholder key to value
+ * @param {object} options options object
+ * @return {string}
+ */
+
+export function translatePlural(app: string, textSingular: string, textPlural: string, number: number, vars?: object, options?: TranslationOptions): string {
+	const identifier = '_' + textSingular + '_::_' + textPlural + '_'
+	const bundle = getAppTranslations(app)
+	const value = bundle.translations[identifier]
+
+	if (typeof (value) !== 'undefined') {
+		const translation = value
+		if (Array.isArray(translation)) {
+			const plural = bundle.pluralFunction(number)
+			return translate(app, translation[plural], vars, number, options)
+		}
+	}
+
+	if (number === 1) {
+		return translate(app, textSingular, vars, number, options)
+	} else {
+		return translate(app, textPlural, vars, number, options)
+	}
+}
+
+/**
+ * Load an app's translation bundle if not loaded already.
+ *
+ * @param {string} appName name of the app
+ * @param {Function} callback callback to be called when
+ * the translations are loaded
+ * @return {Promise} promise
+ */
+export function loadTranslations(appName: string, callback: (...args: any[]) => any) {
+	// already available ?
+	if (hasAppTranslations(appName) || getLocale() === 'en') {
+		const deferred = $.Deferred()
+		const promise = deferred.promise()
+		promise.then(callback)
+		deferred.resolve()
+		return promise
+	}
+
+	const url = generateFilePath(appName, 'l10n', getLocale() + '.json')
```

Comment:
this will call OC methods, thus also breaking the standalone implementation require to [use this library without mocking it](https://github.com/nextcloud/nextcloud-l10n/issues/271#issuecomment-836902307)

Reply:
`generateFilePath` works fine without `OC` being available, as it provides fallbacks for this scenario.

Or are you talking about downloading the file which will fail without an actual NC instance being available?
In this case should we check for `OC` being defined and only then send the request?
Or should be catch 404 and handle 404 as an empty result + emitting a warning?

Reply:
Ah right, I fixed it https://github.com/nextcloud/nextcloud-router/commit/a5fa23f421c51923fd2d478ffa9d0f0c6c398d16